packages/data-provider/src/services/submissionService.ts

```diff
@@ -520,14 +521,18 @@ const service = (dependencies: BaseDependencies) => {
 			userName: string;
 		}): Promise<CreateSubmissionResult> => {
 			logger.info(LOG_MODULE, `Processing '${files.length}' files on category id '${categoryId}'`);
-			const { checkFileNames, checkEntityFieldNames, createOpenActiveSubmission } = submissionUtils(dependencies);
+			const { checkFileNames, checkEntityFieldNames, getOrCreateActiveSubmission } = submissionUtils(dependencies);
 			const { getActiveDictionaryByCategory } = categoryRepository(dependencies);
 
 			const entitiesToProcess: string[] = [];
 			const batchErrors: BatchError[] = [];
+			let status: CreateSubmissionStatus = CREATE_SUBMISSION_STATUS.INVALID_SUBMISSION;
+			let description: string = 'No valid files for submission';
+			let activeSubmissionId: number | undefined = undefined;
```

Comment:
When i see multiple mutable variables and some arrays that will be pushed to I get the sense that we're about to see some complicated code that can maybe be reogranized... I think thats the case here.

The main concern I have is that we have a bunch of default values on these strings that we want to use for the "invalid submission" case. This can be avoided in the flow here by having multiple places where we return from this function. At the end (line 579) we can return an INVALID_SUBMISSION object. The two if statements on lines 570-576 can return PROCESSING or PARTIAL_SUBMISSION. We also have a case where `files.length > 0` and `entitiesToProcess.length === 0` that could be INVALID_SUBMISSION but should maybe have a different message than `'No valid files for submission'`.

I was struggling to explain how to modify it so I rewrote it slightly to illustrate how this could be accomplished without the mutable variables. The advantage here is you can see every case that results in returned data and be sure what the data will be when it is returned from that position:

```ts
		uploadSubmission: async ({
			files,
			categoryId,
			organization,
			userName,
		}: {
			files: Express.Multer.File[];
			categoryId: number;
			organization: string;
			userName: string;
		}): Promise<CreateSubmissionResult> => {
			logger.info(LOG_MODULE, `Processing '${files.length}' files on category id '${categoryId}'`);
			const { checkFileNames, checkEntityFieldNames, getOrCreateActiveSubmission } = submissionUtils(dependencies);
			const { getActiveDictionaryByCategory } = categoryRepository(dependencies);

			if (files.length === 0) {
				return {
					status: CREATE_SUBMISSION_STATUS.INVALID_SUBMISSION,
					description: 'No valid files for submission',
					batchErrors: [],
					inProcessEntities: [],
				};
			}
			const currentDictionary = await getActiveDictionaryByCategory(categoryId);

			if (_.isEmpty(currentDictionary)) {
				throw new BadRequest(`Dictionary in category '${categoryId}' not found`);
			}

			const schemasDictionary: SchemasDictionary = {
				name: currentDictionary.name,
				version: currentDictionary.version,
				schemas: currentDictionary.schemas,
			};

			// step 1 Validation. Validate entity type (filename matches dictionary entities, remove duplicates)
			const schemaNames: string[] = schemasDictionary.schemas.map((item) => item.name);
			const { validFileEntity, batchErrors: fileNamesErrors } = await checkFileNames(files, schemaNames);

			// step 2 Validation. Validate fieldNames (missing required fields based on schema)
			const { checkedEntities, fieldNameErrors } = await checkEntityFieldNames(schemasDictionary, validFileEntity);

			const batchErrors = [...fileNamesErrors, ...fieldNameErrors];
			const entitiesToProcess = Object.keys(checkedEntities);

			if (_.isEmpty(checkedEntities)) {
				return {
					status: CREATE_SUBMISSION_STATUS.INVALID_SUBMISSION,
					description: 'No valid entities in submission',
					batchErrors,
					inProcessEntities: entitiesToProcess,
				};
			}

			// Get Active Submission or Open a new one
			const activeSubmission = await getOrCreateActiveSubmission({ categoryId, userName, organization });
			const activeSubmissionId = activeSubmission.id;

			// Running Schema validation in the background do not need to wait
			// Result of validations will be stored in database
			validateFilesAsync(checkedEntities, {
				categoryId,
				organization,
				userName,
			});

			if (batchErrors.length === 0) {
				return {
					status: CREATE_SUBMISSION_STATUS.PROCESSING,
					description: 'Submission files are being processed',
					submissionId: activeSubmissionId,
					batchErrors,
					inProcessEntities: entitiesToProcess,
				};
			}

			return {
				status: CREATE_SUBMISSION_STATUS.PARTIAL_SUBMISSION,
				description: 'Some Submission files are being processed while others were unable to process',
				submissionId: activeSubmissionId,
				batchErrors,
				inProcessEntities: entitiesToProcess,
			};
		}
```

Reply:
Thanks for the feedback. Refactored this function as suggested.